incubator/validator-avro/src/main/java/io/aklivity/zilla/runtime/validator/avro/AvroReadValidator.java

```diff
@@ -0,0 +1,138 @@
+{
+    public AvroReadValidator(
+        AvroValidatorConfig config,
+        LongFunction<CatalogHandler> supplyCatalog)
+    {
+        super(config, supplyCatalog);
+    }
+
+    @Override
+    public int validate(
+        DirectBuffer data,
+        int index,
+        int length,
+        ValueConsumer next)
+    {
+        return validateComplete(data, index, length, next);
+    }
+
+    @Override
+    public int validate(
+        int flags,
+        DirectBuffer data,
+        int index,
+        int length,
+        FragmentConsumer next)
+    {
+        return (flags & FLAGS_FIN) != 0x00
+            ? validateComplete(data, index, length, (b, i, l) -> next.accept(FLAGS_COMPLETE, b, i, l))
+            : 0;
+    }
+
+    private int validateComplete(
+        DirectBuffer data,
+        int index,
+        int length,
+        ValueConsumer next)
+    {
+        int valLength = -1;
+
+        byte[] payloadBytes = new byte[length];
+        data.getBytes(index, payloadBytes);
+
+        int schemaId;
+        int progress = 0;
+        if (data.getByte(index) == MAGIC_BYTE)
+        {
+            progress += BitUtil.SIZE_OF_BYTE;
+            schemaId = data.getInt(index + progress);
+            progress += BitUtil.SIZE_OF_INT;
+        }
+        else
+        {
+            schemaId = catalog != null ? catalog.id : 0;
+        }
+
+        Schema schema = fetchSchema(schemaId);
+        if (schema != null)
+        {
+            if ("json".equals(format))
+            {
+                byte[] record = deserializeAvroRecord(schema, payloadBytes, progress, length - progress);
+                valueRO.wrap(record);
+                next.accept(valueRO, 0, valLength);
+                valLength = record.length;
+            }
+            else if (validate(schema, payloadBytes, progress, length - progress))
+            {
+                next.accept(data, index, length);
+                valLength = length;
+            }
+        }
+        return valLength;
+    }
+
+    private byte[] deserializeAvroRecord(
+        Schema schema,
+        byte[] bytes,
+        int offset,
+        int length)
+    {
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        try
+        {
+            reader = new GenericDatumReader(schema);
+            GenericRecord record = reader.read(null,
+                decoder.binaryDecoder(bytes, offset, length, null));
+            JsonEncoder jsonEncoder = encoder.jsonEncoder(record.getSchema(), outputStream);
+            writer = record instanceof SpecificRecord ?
+                new SpecificDatumWriter<>(record.getSchema()) :
+                new GenericDatumWriter<>(record.getSchema());
+            writer.write(record, jsonEncoder);
+            jsonEncoder.flush();
+            outputStream.close();
+        }
+        catch (IOException e)
+        {
+        }
+        return outputStream.toByteArray();
```

Comment:
Note that `.toByteArray()` is creating a copy of the underlying array so there is more left to optimize, but this will likely be the last one we sort out.